pkg/ruler/api.go

```diff
@@ -815,3 +828,11 @@ func alertStateDescToPrometheusAlert(d *AlertStateDesc) *Alert {
 
 	return a
 }
+
+// isGCSObjectMutationRateLimitError checks whether the error message is from GCS and is an object mutation rate limit.
+// This is a per-object limit and is limited to 1 mutation per second (https://cloud.google.com/storage/quotas).
+func isGCSObjectMutationRateLimitError(err error) bool {
+	errStr := err.Error()
+	return strings.Contains(errStr, "googleapi: Error 429") &&
+		strings.Contains(errStr, "exceeded the rate limit for object mutation operations")
```

Comment:
Nit: For extra assuredness, would it be worth trying to cast the error into a GCS-specific type? e.g. it seems to be either [this](https://github.com/googleapis/gax-go/blob/cb080069bb4bc31f6598655e94555d9f495f8c1e/v2/apierror/apierror.go#L219) or [this](https://github.com/googleapis/gax-go/blob/cb080069bb4bc31f6598655e94555d9f495f8c1e/v2/apierror/apierror.go#L219), both of which have a way to get more strongly typed details

Completely optional!

Reply:
Updated in https://github.com/grafana/mimir/pull/12086/commits/38acd60d5a5627521c1b57d3d67e67864a83365d to cast to `googleapi.Error` as per https://cloud.google.com/go/docs/reference/cloud.google.com/go/storage/latest#hdr-Errors. I don't think the gax error is used in the case, but if so, [it would wrap the `googleapi.Error`](https://github.com/grafana/mimir/blob/392588b98e386c0fd7cdd549a54e5f06d3155bc2/vendor/github.com/googleapis/gax-go/v2/apierror/apierror.go#L258-L261), so that would still work with `errors.As(err, &gErr)`